lib/chef/resource/macos_userdefaults.rb

```diff
@@ -102,149 +104,63 @@ class MacosUserDefaults < Chef::Resource
         end
 
-        value plist_data[new_resource.key]
-      end
-
-      #
-      # The defaults command to export a domain
-      #
-      # @return [Array] defaults command
-      #
-      def defaults_export_cmd(resource)
-        state_cmd = ["/usr/bin/defaults"]
-
-        if resource.host == "current"
-          state_cmd.concat(["-currentHost"])
-        elsif resource.host # they specified a non-nil value, which is a hostname
-          state_cmd.concat(["-host", resource.host])
-        end
-
-        state_cmd.concat(["export", resource.domain, "-"])
-        state_cmd
+        value state
       end
 
       action :write, description: "Write the value to the specified domain/key." do
         converge_if_changed do
-          cmd = defaults_modify_cmd
-          Chef::Log.debug("Updating defaults value by shelling out: #{cmd.join(" ")}")
-
-          shell_out!(cmd, user: new_resource.user)
+          write_preferences(new_resource)
         end
       end
 
       action :delete, description: "Delete a key from a domain." do
         # if it's not there there's nothing to remove
         return unless current_resource
 
-        converge_by("delete domain:#{new_resource.domain} key:#{new_resource.key}") do
+        # TODO: implement in CF and use here
+      end
 
-          cmd = defaults_modify_cmd
-          Chef::Log.debug("Removing defaults key by shelling out: #{cmd.join(" ")}")
+      action_class do
+        CF_MAPPING = {
+          current_user: CF::Preferences::CURRENT_USER,
+          all_users: CF::Preferences::ALL_USERS,
+          current_host: CF::Preferences::CURRENT_HOST,
+          all_hosts: CF::Preferences::ALL_HOSTS,
+          current: CF::Preferences::CURRENT_HOST # TODO: deprecation warning for this option
+        }
+
+        def read_preferences(new_resource)
+          CF::Preferences.get(new_resource.key, new_resource.domain, mapped_user, mapped_host)
+        end
 
-          shell_out!(cmd, user: new_resource.user)
+        def write_preferences(new_resource)
+          CF::Preferences.set(new_resource.key, new_resource.value, new_resource.domain, mapped_user, mapped_host)
         end
-      end
 
-      action_class do
-        #
-        # The command used to write or delete delete values from domains
-        #
-        # @return [Array] Array representation of defaults command to run
-        #
-        def defaults_modify_cmd
-          cmd = ["/usr/bin/defaults"]
-
-          if new_resource.host == :current
-            cmd.concat(["-currentHost"])
-          elsif new_resource.host # they specified a non-nil value, which is a hostname
-            cmd.concat(["-host", new_resource.host])
-          end
-
-          cmd.concat([action.to_s, new_resource.domain, new_resource.key])
-          cmd.concat(processed_value) if action == :write
-          cmd.prepend("sudo") if new_resource.sudo
-          cmd
+        def mapped_user
+          CF_MAPPING[valid_user.to_sym] || valid_user.to_s
```

Comment:
Is it really intended here that `new_resource.user` could be set to `:all_hosts` which would then map to `CF::Preferences::ALL_HOSTS`?

Reply:
(you should probably write a switch statement to map things like `:all_hosts` onto `CF::Preferences::ALL_HOSTS` and not rely on the rosetta-stone Hash map, or filter the values via a switch statement before grabbing them out of the Hash map).

Reply:
Are we allowed to pass constants like `CF::Preferences::ALL_HOSTS` in recipes? I assumed we only allow simple/built-in ruby objects, hence, tokens like `:all_hosts`

I agree on the switch statement, it's turning out to be more than a 1:1 mapping, thanks 👍

Reply:
oh yeah directly using ruby class constants in cookbook code is terrible tight coupling.  you definitely should use something like `:all_hosts`, my only concern was about the mix+match across the different APIs.